### ISSUE TITLE
Real error no longer propagated through, always -999 cancelled

### ISSUE THREAD

**Issue by alexandre-g** [SDWebImage/SDWebImage#2589]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name        |   ios
 Platform Version     |   11
 SDWebImage Version  |4.4.3
 Integration Method   |   carthage
 Xcode Version           |Xcode 10
 Repro rate              |all the time
 Repro with our demo prj |yes
### Issue Description and Steps

Before upgrading to SDWebImage v4 the code used to relay the real network error to call site.
This would allow to have some logic depending on the error.

Now all "invalid" status codes result in -999 cancelled, which is inaccurate and confusing.

```
SDWebImageManager.shared().loadImage(with: projectPage.timestampedUrl(), options: .retryFailed, progress: { receivedSize, expectedSize, url in

}, completed: { image, data, error, cacheType, finished, imageURL in
    if let error = error {
         /// error is always -999
    }
})
```

**Comment by alexandre-g** [comment]
Still an issue

**Comment by dreampiggy** [comment]
The error code -999, is from the `cancel` method call of download operation. And it should be propagated to the user (Or there are no any callback).

If you don't care about the User or Framework which cancell the download process, you can filter this error code.


```swift
if let error = error, error.code != NSURLErrorCancelled {
  // ...
}
```

**Comment by alexandre-g** [comment]
@dreampiggy I think you misunderstand the issue. What I meant is in the older version of SDWebImage, if the server returned, for example, 403 on image download, SDWebImage would show 403 as the error in the callback. Now, whatever the error is, it's shown as -999 cancelled.
This means we can't have any logic to handle these errors, and, if I remember right, can't even distinguish a cancellation caused by user scrolling an image off screen in a UICollectionView from a real error.

**Comment by dreampiggy** [comment]
A little strange. Is this always reproducible ? I can run a local image server which return 404 to have a check.

**Comment by alexandre-g** [comment]
Yes, I believe it's this code here that generates this behaviour:
```
    BOOL valid = statusCode < 400;
    //'304 Not Modified' is an exceptional one. It should be treated as cancelled if no cache data
    //URLSession current behavior will return 200 status code when the server respond 304 and URLCache hit. But this is not a standard behavior and we just add a check
    if (statusCode == 304 && !self.cachedData) {
        valid = NO;
    }
    
    if (valid) {
        for (SDWebImageDownloaderProgressBlock progressBlock in [self callbacksForKey:kProgressCallbackKey]) {
            progressBlock(0, expected, self.request.URL);
        }
    } else {
        // Status code invalid and marked as cancelled. Do not call `[self.dataTask cancel]` which may mass up URLSession life cycle
        disposition = NSURLSessionResponseCancel;
    }
```

Basically if the status code is > 400, then the task is cancelled

**Comment by zhongwuzw** [comment]
@alexandre-g Thanks for your investigates, it's the bug.

**Comment by dreampiggy** [comment]
@alexandre-g Seems I misunderstand your use case, so please, provide your detailed use case and I can provide solution, or listen to your feature request and work a PR for this.

 ### If you care about the HTTP response status code
In this use case, actually, you actually not care about what `NSError` is or what code it is. You need to check the `NSHTTPURLResponse.statusCode`

In 5.x, you can use the following code to check the status code.

```swift
    func testStatusCode() {
        let imageView = UIImageView()
        let url = URL(string: "https://zh.wikipedia.org/wiki/File:PNG_transparency_demonstration_1.png")!
        imageView.sd_setImage(with: url) { (image, error, cacheType, url) in
            let operation = imageView.sd_imageLoadOperation(forKey: "UIImageView") as! SDWebImageCombinedOperation
            let downloadToken = operation.loaderOperation as! SDWebImageDownloadToken
            let response = downloadToken.response as! HTTPURLResponse
            let statusCode = response.statusCode
            print("status code: \(statusCode)")
        }
    }
```

However, in 4.x (master branch), there are no such public API to check. But you can use NSNotification to add listen to `SDWebImageDownloadReceiveResponseNotification`, which will give you a download operation and you can get the HTTP response. The code looks like below. (Also works for 5.x)

```swift
    func testNotificationStatusCode() {
        NotificationCenter.default.addObserver(forName: .SDWebImageDownloadReceiveResponse, object: nil, queue: nil) { (notification) in
            if let downloadOperation = notification.object as? SDWebImageDownloaderOperation {
                let response = downloadOperation.response as! HTTPURLResponse
                let request = downloadOperation.request!
                let statusCode = response.statusCode
                print("url: \(request.url!.absoluteString)")
                print("status code: \(statusCode)")
            }
        }
    }
```

**Comment by dreampiggy** [comment]
# If you care about the native URLSession error in `NSURLErrorDomain` instead of HTTP standard status code

But, if you care about the native NSError error code listed in `NSURLErrorDomain`. I agree that current implementation, cause the URLSession framework, to always produce a `NSURLErrorCancelled` code, because we implements the delegate method, and return that `NSURLSessionResponseCancel`.

Maybe we can provide a way, to let we pass `NSURLSessionResponseAllow`, but manual cancel the data task. However, this cause a memory leak due to #1744 . This need another workaround.

One solution, is that, we translate some standard HTTP status code, into the native `NSURLErrorDomain` code. Maybe it's not a good idea but a quick fix.

**Comment by dreampiggy** [comment]
This issue, actually already been fixed in version 5.0.

When your HTTP status code is outside than 200-400, such as 403 Forbidden, There is a NSError, with the `SDWebImageErrorInvalidDownloadStatusCode` code and userInfo SDWebImageErrorDownloadStatusCodeKey` represent the HTTP status code.

You can use this to monitor the real HTTP error.

```
NSError *error;
if ([error.domain isEqualToString:SDWebImageErrorDomain] && code == SDWebImageErrorInvalidDownloadStatusCode) {
    int statusCode = [error.userInfo[SDWebImageErrorDownloadStatusCodeKey] intValue];
    switch (statusCode) {
        case 403:
           // forbidden
           break;
        case 404:
           // not found
           break;
        case 500:
           // server error
           break;
    }
}
```
